internal/subscription/delay.go

```diff
@@ -0,0 +1,199 @@
+package subscription
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/razorpay/metro/internal/topic"
+)
+
+// subs-delay-30-seconds, subs-delay-60-seconds ... subs-delay-600-seconds
+const delayTopicNameFormat = "%v-delay-%v-seconds"
+
+const (
+	defaultMinimumBackoffInSeconds    uint = 10
+	lowerBoundMinimumBackoffInSeconds uint = 0
+	upperBoundMinimumBackoffInSeconds uint = 600
+
+	defaultMaximumBackoffInSeconds    uint = 600
+	lowerBoundMaximumBackoffInSeconds uint = 0
+	upperBoundMaximumBackoffInSeconds uint = 3600
+
+	defaultMaxDeliveryAttempts    int32 = 5
+	lowerBoundMaxDeliveryAttempts int32 = 1
+	upperBoundMaxDeliveryAttempts int32 = 100
+)
+
+var (
+	// ErrInvalidMinBackoff ...
+	ErrInvalidMinBackoff = errors.New(fmt.Sprintf("min backoff should be between %v and %v seconds", lowerBoundMinimumBackoffInSeconds, upperBoundMinimumBackoffInSeconds))
+	// ErrInvalidMaxBackoff ...
+	ErrInvalidMaxBackoff = errors.New(fmt.Sprintf("max backoff should be between %v and %v seconds", lowerBoundMaximumBackoffInSeconds, upperBoundMaximumBackoffInSeconds))
+	// ErrInvalidMaxDeliveryAttempt ...
+	ErrInvalidMaxDeliveryAttempt = errors.New(fmt.Sprintf("max delivery attempt should be between %v and %v", lowerBoundMaxDeliveryAttempts, upperBoundMaxDeliveryAttempts))
+)
+
+// Interval is internal delay type per allowed interval
+type Interval uint
+
+var (
+	// Delay5sec 5sec
+	Delay5sec Interval = 5
+	// Delay30sec 30sec
+	Delay30sec Interval = 30
+	// Delay60sec 1min
+	Delay60sec Interval = 60
+	// Delay150sec 2.5min
+	Delay150sec Interval = 150
+	// Delay300sec 5min
+	Delay300sec Interval = 300
+	// Delay600sec 10min
+	Delay600sec Interval = 600
+	// Delay1800sec 30min
+	Delay1800sec Interval = 1800
+	// Delay3600sec 60min
+	Delay3600sec Interval = 3600
+)
+
+var (
+	// MinDelay ...
+	MinDelay = Delay5sec
+	// MaxDelay ...
+	MaxDelay = Delay3600sec
+)
+
+// Intervals during subscription creation, query from the allowed intervals list, and create all the needed topics for retry.
+var Intervals = []Interval{Delay5sec, Delay30sec, Delay60sec, Delay150sec, Delay300sec, Delay600sec, Delay1800sec, Delay3600sec}
+
+// DelayConsumerConfig ...
+type DelayConsumerConfig struct {
+	Topic           string `json:"topic"`
+	GroupID         string `json:"group_id"`
+	GroupInstanceID string `json:"group_instance_id"`
+}
+
+// LogFields ...
+func (dc DelayConsumerConfig) LogFields(kv ...interface{}) []interface{} {
+	fields := []interface{}{
+		"delayConsumerConfig", map[string]interface{}{
+			"topic":           dc.Topic,
+			"groupID":         dc.GroupID,
+			"groupInstanceID": dc.GroupInstanceID,
+		},
+	}
+
+	fields = append(fields, kv...)
+	return fields
+}
+
+// DelayConfig ...
+type DelayConfig struct {
```

Comment:
do we need to duplicate this, we already have retry config and deadletter config in subscritpion.

Reply:
So DelayConfig is a list of all delay topics created for the subscription based on the intervals that were available at that time. Need to keep this as a separate config because if we modify the interval buckets in future, this subscription will remain unaffected and the change will apply only to the new ones.